torch/distributed/_tools/sac_ilp.py

```diff
@@ -103,6 +108,50 @@ def sac_milp(
             ):
                 prob += y[i] == 0
 
+        # Extract FQNs of leaf nodes from the graph
```

Comment:
This is how I would code here (might have syntax errors, but the idea is this)
```suggestion
    node_to_param_size: Dict[str, int] = {
        node["fqn"]: node["param_per_module"] for node in graph.nodes
    }
    
    for fqn in graph["mod_order"]["fw_post_order"]:  # using post order so that children are visited first
        if fqn in fsdp_units:
            node = graph.name2node[fqn]
            j = node["index"]
            # its ancestor nodes will no longer need to take care of these parameters, so subtracting for them
            for i in range(node["index"]):
                if graph.ad_matrix[i][j]:
                    node_to_param_size[graph.nodes[i]["fqn"] -= node_to_param_size[fqn]
    
    # Find maximum parameter count among FSDP units
    max_fsdp_unit_memory = max(
        node_to_param_size[node["fqn"]] * (1 if node["fqn"] in fsdp_units else 0)
        for node in graph.nodes
    )
```